src/find.c

```diff
@@ -30,14 +30,22 @@ rcutils_find(const char * str, char delimiter)
   if (!str || strlen(str) == 0) {
     return 0;
   }
+  return rcutils_findn(str, delimiter, strlen(str));
+}
+
+size_t
+rcutils_findn(const char * str, char delimiter, size_t string_length)
+{
+  if (!str) {
+    return 0;
```

Comment:
I know you didn't come up with this to start with, but I think this is a bug.  Consider the case where the delimiter is the zero'th character.  In that case, the caller can't tell if they got a zero back because the string was NULL, or because the delimiter was in the zeroth position.  I think this should probably return an `ssize_t`, and return a negative value in the `!str` case.

Reply:
We consider that an invalid argument (or at least we do in the string_map https://github.com/ros2/rcutils/blob/42e7840defd2e1150df2f08eb802a4708283376c/src/string_map.c#L238), so if they specify invalid arguments then I think it's ok if they have ambiguous results (if it were an error that happened on our part then I agree), do you think that's fair? I clarified that it'll return 0 on `!str` in the docs though because it was missing: 8aa2d6c1f3bd214a0f0e10475b0c3912fb938298

Reply:
It would be a very tiny performance improvement if this returned `SIZE_MAX` on invalid argument or the delimiter not being found. Calling code could have just one comparison against `SIZE_MAX` if it didn't need the string for anything else. Currently calling code needs to check if the string is valid, then get the length of the string, then after calling this function it must check if the return matches the length of the string it calculated.

I think it makes sense semantically for invalid arguments to return the same thing as a string not being found. It's like saying it didn't find the delimiter because there was no place to look.

Reply:
@sloretz That's a good idea, I like it.  It doesn't change the return type, and it does make semantic sense.  It would require a survey of the existing callers to make sure that they do the correct thing on the changed return value, but I think that would be better.  @dhood Do you agree?  If so, I'd say that that is a big enough change that we should do it in a different PR, and I'll open up an issue about it so that we can move forward on this one (and not forget about it in the future).

Reply:
seems reasonable to me, but as you noted I didn't write this originally, so it might have been returning `size` for a particular reason. I'm sure we can adapt regardless